src/main/java/net/snowflake/client/jdbc/cloud/storage/SnowflakeGCSClient.java

```diff
@@ -1032,14 +1020,19 @@ private void setupGCSClient(
 
     try {
       String accessToken = (String) stage.getCredentials().get("GCS_ACCESS_TOKEN");
-      GoogleCredentials googleCreds;
       if (accessToken != null) {
-        AccessToken googleAccessToken = new AccessToken(accessToken, null);
+        // Using GoogleCredential with access token will cause IllegalStateException when the token
+        // is
+        // expired and trying to refresh, which cause error cannot be caught. Instead, set a header
+        // so we can caught the error code.
 
-        googleCreds = GoogleCredentials.create(googleAccessToken);
         // We are authenticated with an oauth access token.
         this.gcsClient =
-            StorageOptions.newBuilder().setCredentials(googleCreds).build().getService();
+            StorageOptions.newBuilder()
+                .setHeaderProvider(
+                    FixedHeaderProvider.create("Authorization", "Bearer " + accessToken))
+                .build()
+                .getService();
```

Comment:
add a comment as to why this change is made

Reply:
I did add comment above it already

Reply:
ok thanks